core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java

```diff
@@ -0,0 +1,585 @@
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.slice.Slice;
+import io.airlift.units.Duration;
+import io.trino.Session;
+import io.trino.memory.context.LocalMemoryContext;
+import io.trino.operator.OperationTimer.OperationTiming;
+import io.trino.spi.Page;
+import io.trino.spi.PageBuilder;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import io.trino.spi.connector.ConnectorPageSink;
+import io.trino.spi.type.Type;
+import io.trino.split.PageSinkManager;
+import io.trino.sql.planner.PartitioningHandle;
+import io.trino.sql.planner.ScaleWriterPartitioningHandle;
+import io.trino.sql.planner.SystemPartitioningHandle;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.TableWriterNode;
+import io.trino.util.AutoCloseableCloser;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.Futures.allAsList;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.airlift.concurrent.MoreFutures.toListenableFuture;
+import static io.trino.SystemSessionProperties.isStatisticsCpuTimerEnabled;
+import static io.trino.operator.TableWriterOperator.FRAGMENT_CHANNEL;
+import static io.trino.operator.TableWriterOperator.ROW_COUNT_CHANNEL;
+import static io.trino.operator.TableWriterOperator.STATS_START_CHANNEL;
+import static io.trino.operator.TableWriterOperator.TableWriterInfo;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.VarbinaryType.VARBINARY;
+import static io.trino.sql.planner.ScaleWriterPartitioningHandle.unwrapScalingPartitioningHandle;
+import static io.trino.sql.planner.plan.TableWriterNode.CreateTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.InsertTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.RefreshMaterializedViewTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.TableExecuteTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.WriterTarget;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.function.Function.identity;
+
+/**
+ * It partitions the pages and then create a separate pageSink for each partitioned page.
+ * Therefore, it allows us to track physicalWrittenBytes per partition.
+ */
+public class PartitionedTableWriterOperator
```

Comment:
There is a lot of overlap in functionality and a lot of code copied from `TableWriterOperator`. Can this be somehow reused? This could improve the readability of this class by removing non-essential functionality.

Reply:
Consider adding `WriterUtils` class for shared code